### ISSUE TITLE
What to do with this library ?

### ISSUE THREAD

**Issue by antony-jr** [antony-jr/AppImageUpdaterBridge#6]
As discussed in #1  , AppImageUpdaterBridge is now in pure C++ with a very good Qt API. I even re-implemented Zsync(As discussed) in C++ for it work more efficiently(And it does work good). To the best of my knowledge the code base is clean using only Qt API's and ensures portability.

### These are the things I did 

* Re-implemented Zsync in C++ inheriting only the hash table.
> zsync2 has a C++11 interface, so, standard types instead of the Qt versions (but actually it's all compatible...). I don't see the issue leading to the decision of @antony-jr to rewrite the original zsync (which has had a lot of issues, which took many weeks of intensive work to fix all those issues in zsync2. You shouldn't underestimate the amount of engineering that was necessary to create both libzsync2 and libappimageupdate. -- TheAssassin

* Re-implemented the whole library in single threaded non-blocking API using Qt's event loop , Following a Worker-Object model.(Strictly , even for Zsync stuff)

* Fully replaced libcurl's network support with Qt's Network Support , Now this library is generally faster when it comes to network since QNetworkAccessManager can do concurrent http request (maximum 6 at a time). 
>  QNetworkAccessManager queues the requests it receives. The number of requests executed in parallel is dependent on the protocol. Currently, for the HTTP protocol on desktop platforms, 6 requests are executed in parallel for one host/port combination.

* Added a **AppImageUpdaterDialog** class(not documented yet) which inherits QDialog and has the delta updater intrinsic to it  , thus , once ```init()``` is called , Idle timer is started , The user can set the idle seconds , The default idle seconds is 0 seconds which will make the timer timeout when the event loop is free (i.e When the application is not doing anything). But the user can also override the *event* virtual method from ```QWidget``` or ```QMainWindow``` to reset the idle timer whenever an event is triggered to find the best time to show the self updater.(i.e show the updater when there is no mouse or keyboard events to their application)

* Intelligent guessing of self update , i.e If you construct a *AppImageDeltaRevisioner* (Thats the main updater class) without any AppImage path , It will automatically guess the current AppImage path that is running the updater and so does not expect the user to provide any path , If you want to do a self update , Simply construct the updater class and start it off , Same goes for **AppImageUpdaterDialog**.

* Completely thread safe, Tested with hardcore threading tests , But there is still no good way to test threads.(Thanks to Qt Signals and Slots.)

* If HTTP Partial content is not supported by the server then automatically falls back to full download without doing the delta update , In this state , The updater simply becomes a traditional http downloader.

* Does not do segmentation violation in any case to the best of my knowledge.

* Cancel , Start and other control methods works smoothly.

* Explicit logging if the developer wants , Disabling the logging can reduce overhead and binary size.

* Does not touch the host system , like AppImageUpdate does(i.e AppImageUpdate sometimes deletes user files without any content integrity checks but this library does not touch the host machine at all , if the target file name exists it will rename the new target file with the current date time in [ISO 8601](https://www.iso.org/standard/40874.html) format placed in the middle of the filename , Thus avoiding deleting any files in the host machine.) 

Try the new [AppImageUpdater](https://github.com/antony-jr/AppImageUpdater) to test these out in action.

Here are some screenshots from AppImageUpdater ( You can get an idea on how fluid it is. )

**Step 1**: Start AppImageUpdater's latest version. ( It should auto update if any new releases are found).

![1](https://s15.postimg.cc/8gcfxvnmx/image.png?dl=1)

**Step 2**: Run it.

![2](https://s15.postimg.cc/fwbpjoj21/image.png?dl=1)

![3](https://s15.postimg.cc/axo75akft/image.png?dl=1)

![4](https://s15.postimg.cc/hbda8kch5/image.png?dl=1)

![5](https://s15.postimg.cc/yonknescp/image.png?dl=1)

**Step 3**: Drag and drop AppImage(s) to start update.

![6](https://s15.postimg.cc/4jz422puh/image.png?dl=1)

![7](https://s15.postimg.cc/e4iqozhrd/image.png?dl=1)

![8](https://s15.postimg.cc/5m9akmo3d/image.png?dl=1)

![9](https://s15.postimg.cc/p43y0l0gp/image.png?dl=1)

**Step 4**: Directly open your new version of AppImage , AppImageUpdater can be minimized to tray.

![10](https://s15.postimg.cc/u2rgf3h49/image.png?dl=1)

**Step 5**: Use your new version.

![11](https://s15.postimg.cc/w7btg5g61/image.png?dl=1)

**Now the question is , What do we plan to do with this library ? I really need some pointers on what I need to work on or just drop the library in favor of the official library.** 

**CC**: @probonopd , @TheAssassin

**Comment by probonopd** [comment]
Cool @antony-jr - is it also possible to drag-and-drop an entire folder/directory (tree) and have the tool update all AppImages therein recursively?

I am sure this would be a "killer feature" for many.

**Comment by antony-jr** [comment]
@probonopd I think you are missing the point , I asked what do we do with this library , I mean @TheAssassin seems to think this as reinventing the wheel so I suppose we could purge this library in favor of the official library.(Aka. I don't want to waste my time on reinventions but I don't think this is reinventing the wheel)

**And the feature you asked will be publish in 10 minutes from the time of this comment(Currently building in travis-ci)(EDIT: Now published successfully) , The Application already supported multiple AppImages and can now handle both directories and also desktop files(this is added to the core of the library , i.e to parse AppImage path from .desktop files.)**

**EDIT** : It should auto update itself (i.e The AppImageUpdater Application. )
Here are some screenshots on Auto Update(You can also see that the icon pixmap is customizable , Thanks to this library),
![8](https://s33.postimg.cc/iid5t9wp9/image.png?dl=1)

![9](https://s33.postimg.cc/rq5e9zbh9/image.png?dl=1)

Here are some screenshots from my local build, 

As you can see the logger is explicit as you hoped. (logger is printing in the terminal)
![1](https://s33.postimg.cc/dfr8kosvh/image.png?dl=1) 

You can drop any kind of directory(soft links or hard links also) ,
![2](https://s33.postimg.cc/3wm54h2bh/image.png?dl=1)

![3](https://s33.postimg.cc/cevl8u3p9/image.png?dl=1)

You can again confirm with the logger,
![4](https://s33.postimg.cc/85qv6mq59/image.png?dl=1)

It also notifies the user if they are using the new version already,
![5](https://s33.postimg.cc/8v9nj0t9p/image.png?dl=1)

![6](https://s33.postimg.cc/aab87pk25/image.png?dl=1)

You can also see that it knows when the server does not support http partial content and it is noted in the logger. ( The screenshot did not capture well , sorry about that )
![7](https://s33.postimg.cc/ogqz2ypsd/image.png?dl=1)

**Comment by TheAssassin** [comment]
@antony-jr again: I don't want you to drop any code or anything, but we need an upstream library that does not depend on Qt. This way, we can provide developers a library that doesn't suffer from some "political" issues (like "we prefer Gtk+") or even technological issues ("our UI is written in Gtk+/FLTK/..., linking to Qt is not an option", etc.). AppImageUpdate is, after all, somewhat complex code-wise, and there is no need for more than one known-good implementation.

"Reinvent the wheel", well, of course replacing something that works is a bit of that, but if one does want to do it anyway, why not? It's your free decision what to do.

You're welcome to partially rewrite your code base using standard types (C++11), and we could merge it into AppImageUpdate. The UI itself looks interesting, too, and might be merged as well, after having rewritten the core library (I am not a UI designer, and I don't mind replacing my work). C++11 has some async support, and I could imagine there's some Qt interface that allows for using those with Qt's event loop, or something like that.

Please understand that AppImageUpdate is not a priority right now for us (especially UI wise), if you want to help with development, please be our guest. I can compile a list of requirements we have at any time.

I have some time next week to discuss such a project with you. Please visit us on IRC to discuss my suggestion:
 - (completely) rewrite zsync2's C code in C++ (half of that should already be available in your codebase)
 - provide `std::async` based interface for zsync2
 - incorporate changes into AppImageUpdate
 - rewrite libappimageupdate to use async model
 - adapt changes in libappimageupdate-qt, turning it into a complete Qt-style wrapper for libappimageupdate
 - implement new "standard widget" which can be used by existing projects in libappimageupdate-qt

**Comment by antony-jr** [comment]
@TheAssassin The thing is that its really hard to develop Qt Apps when you don't use Qt types , thats reason why I started this project in the first place. Anyways lets talk about the implementation next week , I'm thinking about using **boost** since its close to the standards , don't you think ? So lets assume this as an unofficial version of AppImageUpdate library for Qt Apps. And also I don't like Qt that much , Qt sometimes reminds me of Java and I hate Java so much. Thank You for your time , sorry to disturb.

**Comment by probonopd** [comment]
> What to do with this library

It seems like @antony-jr knows how to do things "the Qt way", while I and (to a lesser degree @TheAssassin) are still relatively new to that and may not do things "the Qt way" yet. @TheAssassin is writing code that can be used outside of Qt, too.

All of this is open source, so why not have both options.
